src/storage/sqlite.js

```diff
@@ -0,0 +1,180 @@
+// @flow strict-local
+import {
+  openDatabase,
+  type SQLResultSet,
+  type WebSQLDatabase,
+  type SQLTransaction as WebSQLTransaction,
+} from 'expo-sqlite';
+import invariant from 'invariant';
+
+/* eslint-disable no-use-before-define */
+/* eslint-disable no-underscore-dangle */
+/* eslint-disable no-void */
+/* eslint-disable flowtype/type-id-match */
+
+type SQLArgument = number | string;
+
+/**
+ * A Promise-based wrapper on expo-sqlite.
+ */
+export class SQLDatabase {
+  db: WebSQLDatabase;
+
+  constructor(name: string) {
+    this.db = openDatabase(name);
+  }
+
+  /**
+   * Like the `transaction` method in expo-sqlite, but promisified.
+   *
+   * The transaction will stay open as long as the Promise the callback
+   * returned stays unsettled.  That means the callback can use `await` as
+   * usual.
+   *
+   * For performance reasons, though, avoid using `await` other than on
+   * `executeSql` results or already-settled values.  In particular, avoid
+   * awaiting network requests or other things that take time but don't need
+   * to occupy the CPU.  That's because the current implementation of this
+   * method effectively busy-waits, by making trivial SQL queries, when SQL
+   * queries are getting to run but the transaction has no other SQL queries
+   * outstanding.
+   */
+  // TODO: It'd be good to be able to remove the performance warning above.
+  //   (Well, part of it; having a transaction block on something
+  //   potentially really slow like a network request will generally be a
+  //   bad idea just because of blocking other things in the app from
+  //   proceeding, even if it doesn't involve busy-waiting.)
+  //
+  //   This will involve modifying expo-sqlite, replacing much of the
+  //   node-websql code under it.
+  //
+  //   More background on that performance warning:
+  //     https://github.com/zulip/zulip-mobile/pull/5184#discussion_r779300110
+  transaction(cb: SQLTransaction => void | Promise<void>): Promise<void> {
+    return new Promise((resolve, reject) =>
+      this.db.transaction(
+        tx => void keepQueueLiveWhile(tx, () => cb(new SQLTransactionImpl(this, tx))),
+        reject,
+        resolve,
+      ),
+    );
+  }
+
+  /**
+   * Like the `readTransaction` method in expo-sqlite, but promisified.
+   *
+   * See `transaction` for performance considerations on how to use `await`
+   * inside the callback.
+   */
+  readTransaction(cb: SQLTransaction => void | Promise<void>): Promise<void> {
+    return new Promise((resolve, reject) =>
+      this.db.readTransaction(
+        tx => void keepQueueLiveWhile(tx, () => cb(new SQLTransactionImpl(this, tx))),
+        reject,
+        resolve,
+      ),
+    );
+  }
+
+  /**
+   * Convenience method for a single read-only query.
```

Comment:
I think this should say that the query is run in its own transaction.

I think we want to design our wrapper not to expose a way to execute SQL statements outside of a transaction, right? That would solve a potential problem Anders points out [on CZO](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/sqlite.20.23M4841/near/1301069):

> […] you could get incorrect interleaving if the external event triggers a non-transactional query during an asynchronous action inside a transaction?

(It solves it because it prevents us from ever triggering non-transactional queries anywhere.)

Reply:
Sure, good thought.